Support GRANT .. ON SCHEMA .. in Babelfish

shalinilohia50 opened the pull request:
### Description
Support GRANT/REVOKE .. ON SCHEMA .. in Babelfish

**PLEASE NOTE:** This change is with the assumption that we are going to restrict function overloading. Based on any other  decision made otherwise, the catalog implementation would be extended to store the arguments in another PR.

1. Supported syntax `GRANT <permission> ON SCHEMA::<schema_name> TO <user_name>`
2. Supported syntax `REVOKE <permission> ON SCHEMA::<schema_name> FROM <user_name>`
3. Added one SQL statement `PLTSQL_STMT_GRANTSCHEMA` to store relevant information to execute GRANT/REVOKE .. ON SCHEMA .. statements.
4. Created one catalog table `sys.babelfish_schema_permissions` to hold the details for each GRANT/REVOKE statements.
5. GRANT on schema/objects adds a row in the catalog table if not exists already.
6. REVOKE on schema/objects removes the corresponding row in the catalog table if it exists already.
7. REVOKE on schema internally grants permission to all the objects if there are explicit permissions granted to the objects belonging to the same schema.
8. GRANT ALL on objects work as it is and add rows in the catalog for each relevant permission depending on the object type.
9. REVOKE ALL on object should do nothing, if the relevant schema permission exists in the catalog. But, it should remove the rows from the catalog if the object level permission is granted.
10. Drop statement for OBJECT/SCHEMA removes all the relevant object entries from the catalog. 

### Issues Resolved

Task: [BABEL-4344](https://jira.rds.a2z.com/browse/BABEL-4344)
Signed-off-by: Shalini Lohia <lshalini@amazon.com>

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** N/A

* **Arbitrary inputs -** N/A

* **Negative test cases -** Yes

* **Minor version upgrade tests -** N/A

* **Major version upgrade tests -** N/A

* **Performance tests -** N/A

* **Tooling impact -** N/A

* **Client tests -** N/A

lohia-shalini committed:
Support GRANT .. ON SCHEMA .. in Babelfish

lohia-shalini committed:
Merge branch 'BABEL_3_X_DEV' into grant-on-schema

lohia-shalini committed:
Fix test failure

lohia-shalini committed:
Fix tests

lohia-shalini committed:
Minor fixes

KushaalShroff commented:
Looks like I have reviewed most of the code in few of the previous PRs already.
1. We have lost all the history from before, can you link all of the PRs so that we can check what is changed.
2. We have lost all the history of comments from reviewers, can we validate if those are addressed?

shalinilohia50 commented:
https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2020

shalinilohia50 commented:
> Looks like I have reviewed most of the code in few of the previous PRs already.
> 
> 1. We have lost all the history from before, can you link all of the PRs so that we can check what is changed.
> 2. We have lost all the history of comments from reviewers, can we validate if those are addressed?

https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2020

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+	int16	dbid = get_cur_db_id();
+
+	/* Immediately return, if grantee is NULL or PUBLIC. */
+	if ((grantee == NULL) || (strcmp(grantee, "public") == 0))
```
strncmp should be used?

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
```
This strategy is backed by nameeqfast and it ignores collations. So does you changes works fine with mixed cases? case insensitively?

shalinilohia50 replied:
Yes, it does. We have tests for that.

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+ * Removes a row from the catalog BABELFISH_SCHEMA_PERMISSIONS.
+ */
+void
+remove_entry_from_bbf_schema_perms(const char *schema_name,
```
will this be called for any object being dropped to remove the entry? how are we ensuring consistency of this table

shalinilohia50 replied:
Yes, it will be called at the time of DROP statement to remove all relevant entries.
We need to add some consistency check on this table. IMO let's handle it separately since it's already a lengthy PR.

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+ * Updates the permission column for a particular row in BABELFISH_SCHEMA_PERMISSIONS table.
+ */
+void
+update_privileges_of_object(const char *schema_name,
```
will this be called for any ALTER statement on the objects?

shalinilohia50 replied:
What do you mean by ALTER statement? 
Permissions can only be granted/revoked using GRANT/REVOKE statements and then we call this function.

Deepesh125 replied:
Not in terms of permission. There can be ALTER object ... RENAME ... or sp_rename execution on objects. We need to ensure that it works with all such operations

shalinilohia50 replied:
Yes, that will be handled separately in another PR. Removing few things out of this PR since this is already a big one.

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+				true, NULL, 5, scanKey);
+	tuple_bbf_schema = systable_getnext(scan);
+
+	if (HeapTupleIsValid(tuple_bbf_schema))
```
what if there is no entry found in the catalog? should it return error?

shalinilohia50 replied:
There will be an entry for an object if we are calling this function. It is ensured here in this function call `add_or_update_object_in_bbf_schema`

KushaalShroff replied:
Assumption based on propriety code shall not be made. This is not extensible and reusable then.

We MUST throw errors everywhere in such Utilities.

shalinilohia50 replied:
This will return 0 in case no entry is found.

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+
+	if (current_permission == 0)
+	{
+		remove_entry_from_bbf_schema_perms(schema_name, object_name, grantee, object_type);
```
As per current code, get_privilege_of_object will return old_permission = 0 even if theres no entry in the catalog and we are trying to remove the entry here. should write more defensive code

shalinilohia50 replied:
There will be an entry for an object if we are calling this function. It is ensured here in this function call `add_or_update_object_in_bbf_schema`

Deepesh125 commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -14,6 +14,17 @@ CREATE TABLE sys.babelfish_sysdatabases (
+  object_name NAME NOT NULL COLLATE sys.database_default,
+  permission INT NOT NULL,
+  grantee NAME NOT NULL COLLATE sys.database_default,
+  object_type NAME COLLATE sys.database_default,
```
object_type can be NULL? should we maintain enum to store object_type instead of string?

shalinilohia50 replied:
No, `object_type` cannot be NULL.

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.h:
```diff
@@ -280,6 +280,61 @@ typedef struct FormData_bbf_function_ext
+#define Anum_bbf_schema_perms_grantee 5
+#define Anum_bbf_schema_perms_object_type 6
+
+#define PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA "ALL"
```
can there be object named "ALL"?

shalinilohia50 replied:
No, it's not allowed.

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+			if (permission == (current_permission & permission))
+			{
+				priv_name = privilege_to_string(permission);
+				if (strcmp(object_type, "r") == 0)
+					appendStringInfo(&query, "GRANT \"%s\" ON \"%s\".\"%s\" TO \"%s\"; ", priv_name, schema, object_name, grantee);
+				else if (strcmp(object_type, "f") == 0)
+					appendStringInfo(&query, "GRANT \"%s\" ON FUNCTION \"%s\".\"%s\" TO \"%s\"; ", priv_name, schema, object_name, grantee);
+				else if (strcmp(object_type, "p") == 0)
+					appendStringInfo(&query, "GRANT \"%s\" ON PROCEDURE \"%s\".\"%s\" TO \"%s\"; ", priv_name, schema, object_name, grantee);
```
1. how do we validate all these params?
2. Better to use macro here instead of "r", "p" etc

shalinilohia50 replied:
1. There is no validation performed here. Based on the object type, we create the grant statement using the object name in the catalog. 
We rely on the object entry in the catalog. If an object is deleted, we anyway delete the respective rows. 
2. Yes.

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+	{
+		Form_bbf_schema_perms schemaform;
+		schemaform = (Form_bbf_schema_perms) GETSTRUCT(tuple_bbf_schema);
+		object_name = pstrdup(NameStr(schemaform->object_name));
+		object_type = pstrdup(NameStr(schemaform->object_type));
```
free?

shalinilohia50 replied:
Cannot free these since these are const char*.

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+							RowExclusiveLock);
+
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(schema_name));
+	ScanKeyEntryInitialize(&scanKey[2], 0,
+				Anum_bbf_schema_perms_object_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(object_name));
+	ScanKeyInit(&scanKey[3],
+				Anum_bbf_schema_perms_permission,
+				BTEqualStrategyNumber, F_INT4EQ,
+				Int32GetDatum(old_permission));
+	ScanKeyEntryInitialize(&scanKey[4], 0,
+				Anum_bbf_schema_perms_grantee,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(grantee));
```
scankeyinit for individual field is almost common everywhere, can we try to create macro to do that?

shalinilohia50 replied:
IMO we should keep this the same way. This follows the standard format we have been doing for all the catalogs. Any opinion?

Deepesh125 commented:
We should have tests cases which involves schema name, object name and grantee name with letters from different languages, regular identifier, quoted identifiers, name length greater than 64, etc.

Deepesh125 commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3687,6 +3688,101 @@ get_insert_bulk_kilobytes_per_batch()
+	}
+	else
+	{
+		ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_SCHEMA),
+					 errmsg("An object or column name is missing or empty. For SELECT INTO statements, verify each column has a name. For other statements, look for empty alias names. Aliases defined as \"\" or [] are not allowed. Change the alias to a valid name.")));
```
do we have test cases for this error?

shalinilohia50 replied:
Yes, we have.

Deepesh125 commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3687,6 +3688,101 @@ get_insert_bulk_kilobytes_per_batch()
+					 errmsg("An object or column name is missing or empty. For SELECT INTO statements, verify each column has a name. For other statements, look for empty alias names. Aliases defined as \"\" or [] are not allowed. Change the alias to a valid name.")));
+	}
+
+	if (!OidIsValid(schemaOid))
+			ereport(ERROR,
+					(errcode(ERRCODE_UNDEFINED_SCHEMA),
+					 errmsg("schema \"%s\" does not exist",
+							schema_name)));
```
I think we should return logical schema name that will make more sense to customer

Deepesh125 commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3687,6 +3688,101 @@ get_insert_bulk_kilobytes_per_batch()
+		{
+			char	*grantee_name = (char *) lfirst(lc);
+			Oid	role_oid;
+			bool	grantee_is_db_owner = 0 == strncmp(grantee_name, get_owner_of_db(dbname), NAMEDATALEN);
```
can we define one variable outside of this loop to hold get_owner_of_db?

Deepesh125 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1033,16 +1033,49 @@ update_GrantStmt(Node *n, const char *object, const char *obj_schema, const char
+		if (object)
+		{
+			DefElem *tmp = (DefElem *) lfirst(lc);
+			tmp->defname = pstrdup("schemas");
```
should we free tmp->defm=name before allocating new memory?

shalinilohia50 replied:
I verified and we don't need to pfree() this pointer explicitly since it's already NULL at this point.

Deepesh125 commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1033,16 +1033,49 @@ update_GrantStmt(Node *n, const char *object, const char *obj_schema, const char
+
+	if (priv && stmt->privileges)
+	{
+		AccessPriv *tmp = (AccessPriv *) llast(stmt->privileges);
+		tmp->priv_name = pstrdup(priv);
```
1. stmt->privileges will always be non-NIL?
2. tmp->priv_name should be freed before allocating new memory

shalinilohia50 replied:
1. Yes.
2. Are you suggesting this?
```
AccessPriv *tmp = (AccessPriv *) llast(stmt->privileges);
pfree(tmp->priv_name);
tmp->priv_name = pstrdup(priv);
```

KushaalShroff replied:
2. Yes!

shalinilohia50 replied:
I verified and we don't need to pfree() this pointer explicitly since it's already NULL at this point.

shalinilohia50 committed:
Merge branch 'BABEL_3_X_DEV' into grant-on-schema

lohia-shalini committed:
Fix minor comments

KushaalShroff commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -14,6 +14,19 @@ CREATE TABLE sys.babelfish_sysdatabases (
+-- Please avoid using this catalog anywhere else.
+CREATE TABLE sys.babelfish_schema_permissions (
+  dbid smallint NOT NULL,
+  schema_name NAME NOT NULL COLLATE sys.database_default,
```
We should not be using NAME datatype anymore

KushaalShroff commented on contrib/babelfishpg_tsql/sql/ownership.sql:
```diff
@@ -14,6 +14,19 @@ CREATE TABLE sys.babelfish_sysdatabases (
+CREATE TABLE sys.babelfish_schema_permissions (
+  dbid smallint NOT NULL,
+  schema_name NAME NOT NULL COLLATE sys.database_default,
+  object_name NAME NOT NULL COLLATE sys.database_default,
+  permission INT NOT NULL,
+  grantee NAME NOT NULL COLLATE sys.database_default,
+  object_type NAME NOT NULL COLLATE sys.database_default,
```
Lets change the name datatype to sys.n/varchar

Deepesh125 replied:
I and Shalini touched upon this thing offline, in fact we should avoid using NAME data type as it is having some peculiarity for postgres. First and foremost is that, it will always do case insensitive comparison. We understand that this catalog would not be used by customers directly but we may still consult this catalog during our operations/debugging. So lets use the data type appropriately.

shalinilohia50 replied:
Sure. Will change all the NAME datatype to `sys.nvarchar`.

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -1435,6 +1441,29 @@ clean_up_bbf_function_ext(int16 dbid)
+ *****************************************/
+
+Oid
+get_bbf_schema_perms_oid()
```
Should be a static function

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -1435,6 +1441,29 @@ clean_up_bbf_function_ext(int16 dbid)
+}
+
+Oid
+get_bbf_schema_perms_idx_oid()
```
Should be a static function

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+	int16	dbid = get_cur_db_id();
+
+	/* Immediately return, if grantee is NULL or PUBLIC. */
+	if ((grantee == NULL) || (strncmp(grantee, "public", NAMEDATALEN) == 0))
```
Why are we using NAMEDATALEN as the max compare length? This is wrong!

Also We should not unnecessary complicate things,  by using strncmp - we have one the strings as fixed length and not a variable one so strcmp is enough.

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.h:
```diff
@@ -280,6 +280,61 @@ typedef struct FormData_bbf_function_ext
+ *****************************************/
+#define BBF_SCHEMA_PERMS_TABLE_NAME "babelfish_schema_permissions"
+#define BBF_SCHEMA_PERMS_IDX_NAME "babelfish_schema_permissions_pkey"
+#define BBF_SCHEMA_PERMS_NUM_OF_COLS 6
+#define BBF_SCHEMA_PERMS_DBID 0
+#define BBF_SCHEMA_PERMS_SCHEMA_NAME 1
+#define BBF_SCHEMA_PERMS_OBJECT_NAME 2
+#define BBF_SCHEMA_PERMS_PERMISSION 3
+#define BBF_SCHEMA_PERMS_GRANTEE 4
+#define BBF_SCHEMA_PERMS_OBJECT_TYPE 5
+#define Anum_bbf_schema_perms_dbid 1
+#define Anum_bbf_schema_perms_schema_name 2
+#define Anum_bbf_schema_perms_object_name 3
+#define Anum_bbf_schema_perms_permission 4
+#define Anum_bbf_schema_perms_grantee 5
+#define Anum_bbf_schema_perms_object_type 6
```
Explain why do we need duplicate macros. 1 for index starting from 0 and other one from 1. Seems redundant, aint it?

shalinilohia50 replied:
Though there are duplicate macros but they mean differently. for example, BBF_SCHEMA_PERMS_DBID  is 0 while Anum_bbf_schema_perms_dbid is 1. We would need both to make it understandable.

KushaalShroff replied:
Is it too difficult to fetch one from the other?

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+				wrapper->stmt_len = 1;
+
+				/* do this step */
+				ProcessUtility(wrapper,
```
As raised earlier during the reviews, I do not favour the use of ProcessUtility directly. We have carried on with this maybe with other features making it a trend which is not the right solution at hand always.

Can we use SPI instead?

shalinilohia50 replied:
https://jira.rds.a2z.com/browse/BABEL-4698 will be tracked here since it needs more understanding

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+					appendStringInfo(&query, "GRANT \"%s\" ON FUNCTION \"%s\".\"%s\" TO \"%s\"; ", priv_name, schema, object_name, grantee);
+				else if (strcmp(object_type, OBJ_PROCEDURE) == 0)
+					appendStringInfo(&query, "GRANT \"%s\" ON PROCEDURE \"%s\".\"%s\" TO \"%s\"; ", priv_name, schema, object_name, grantee);
+				res = raw_parser(query.data, RAW_PARSE_DEFAULT);
+				res_stmt = ((RawStmt *) linitial(res))->stmt;
+
+				/* need to make a wrapper PlannedStmt */
+				wrapper = makeNode(PlannedStmt);
+				wrapper->commandType = CMD_UTILITY;
+				wrapper->canSetTag = false;
+				wrapper->utilityStmt = res_stmt;
+				wrapper->stmt_location = 0;
+				wrapper->stmt_len = 1;
+
```
If we use SPI, we wont require to call raw_parser ourselves

shalinilohia50 replied:
https://jira.rds.a2z.com/browse/BABEL-4698 will be tracked here since it needs more understanding

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.h:
```diff
@@ -280,6 +280,61 @@ typedef struct FormData_bbf_function_ext
+
+#define PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA "ALL"
+#define ALL_PERMISSIONS_ON_RELATION 47
+#define ALL_PERMISSIONS_ON_FUNCTION 128
```
Add comment with bit positions

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.h:
```diff
@@ -280,6 +280,61 @@ typedef struct FormData_bbf_function_ext
+#define PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA "ALL"
+#define ALL_PERMISSIONS_ON_RELATION 47
+#define ALL_PERMISSIONS_ON_FUNCTION 128
+#define OBJ_SCHEMA "s"
+#define OBJ_RELATION "r"
+#define OBJ_PROCEDURE "p"
+#define OBJ_FUNCTION "f"
```
Any reason why these are chars? are we doing strcmp to read them? Will using integers benefit?

shalinilohia50 replied:
We used NAME datatype initially. Changed this `type` column in the catalog as `char(1)`. 
Any reason to use integers?

KushaalShroff replied:
In fact Should be an enum

shalinilohia50 replied:
These are internal catalog tables. Similar catalog tables like `babelfish_authid_user_ext` and `babelfish_authid_login_ext` use type as `CHAR(1)`. Is CHAR(1) really a problem here?

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.h:
```diff
@@ -280,6 +280,61 @@ typedef struct FormData_bbf_function_ext
+	NameData	object_name;
+	int32		permission;
+	NameData	grantee;
+	NameData	object_type;
```
Why are we storing with Name datatype? Use char* instead (n/varchar).

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+							NULL);
+
+				/* make sure later steps can see the object created here */
+				CommandCounterIncrement();
```
From previous PRs: Why do we need to call CommandCounterIncrement for this?

Please make me recollect if there was any explanation on to why this has not been changed

KushaalShroff commented on contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp:
```diff
@@ -1763,20 +1763,23 @@ void TsqlUnsupportedFeatureHandlerImpl::checkSupportedGrantStmt(TSqlParser::Gran
 				continue;
 			else
 			{
-				unsupported_feature = "GRANT PERMISSION " + perm->getText();
+				unsupported_feature = "GRANT PERMISSION " + ::getFullText(single_perm);
+				std::transform(unsupported_feature.begin(), unsupported_feature.end(), unsupported_feature.begin(), ::toupper);
 				handle(INSTR_UNSUPPORTED_TSQL_REVOKE_STMT, unsupported_feature.c_str(), getLineAndPos(perm));
```
Which are the cases which are not included in the above if condition?

shalinilohia50 replied:
There are many permissions mentioned [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/BABEL_3_X_DEV/contrib/babelfishpg_tsql/antlr/TSqlParser.g4#L2890-L2917) .
We only support few now. 
```
EXECUTE, EXEC, SELECT, INSERT, UPDATE, DELETE, REFERENCES, CONNECT
```

KushaalShroff commented on contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp:
```diff
@@ -1763,20 +1763,23 @@ void TsqlUnsupportedFeatureHandlerImpl::checkSupportedGrantStmt(TSqlParser::Gran
 		auto obj_type = perm_obj->object_type();
-		if (obj_type && !obj_type->OBJECT())
+		if (grant->ALL() && obj_type && obj_type->SCHEMA())
+			throw PGErrorWrapperException(ERROR, ERRCODE_FEATURE_NOT_SUPPORTED, "The all permission has been deprecated and is not available for this class of entity.", getLineAndPos(grant));
```
Grammar of error message doesnt seem right

shalinilohia50 replied:
This is the same error message which SQL Server throws.

KushaalShroff commented on contrib/babelfishpg_tsql/src/tsqlUnsupportedFeatureHandler.cpp:
```diff
@@ -1856,20 +1859,23 @@ void TsqlUnsupportedFeatureHandlerImpl::checkSupportedRevokeStmt(TSqlParser::Rev
 			else
 			{
-				unsupported_feature = "REVOKE PERMISSION " + perm->getText();
+				unsupported_feature = "REVOKE PERMISSION " + ::getFullText(single_perm);
+				std::transform(unsupported_feature.begin(), unsupported_feature.end(), unsupported_feature.begin(), ::toupper);
```
Was this not being rewritten before?

shalinilohia50 replied:
No.

KushaalShroff commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -1919,6 +1919,52 @@ class tsqlBuilder : public tsqlCommonMutator
 			}
 		}
+		else if (ctx->grant_statement() && ctx->grant_statement()->ON() && ctx->grant_statement()->permission_object()
+				&& ctx->grant_statement()->permission_object()->object_type() && ctx->grant_statement()->permission_object()->object_type()->SCHEMA())
```
Lets simplify the if condition by using a temp variable to store ctx->grant_statement() promoting readability.

Also we have been using ctx->grant_statement() inside the block as well

KushaalShroff commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -1919,6 +1919,52 @@ class tsqlBuilder : public tsqlCommonMutator
+						|| single_perm->DELETE()
+						|| single_perm->REFERENCES())
+					{
+						clear_rewritten_query_fragment();
```
WHY??

shalinilohia50 replied:
```
                                                if (single_perm->EXECUTE()
						|| single_perm->EXEC()
						|| single_perm->SELECT() 
						|| single_perm->INSERT()
						|| single_perm->UPDATE()
						|| single_perm->DELETE()
						|| single_perm->REFERENCES())
```
We are only supporting these privileges for this GRANT on SCHEMA syntax. That's why we need this change in `exitSecurity_statement()`

KushaalShroff replied:
The question was why do we need to call clear_rewritten_query_fragment. EXPLAIN.

shalinilohia50 replied:
Oh! I was following [this](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/0a3b2dca475894779afb8ebe302ba21bd411aae3/contrib/babelfishpg_tsql/src/tsqlIface.cpp#L205-L226). 
I think it's not needed.

KushaalShroff commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -1919,6 +1919,52 @@ class tsqlBuilder : public tsqlCommonMutator
+			}
+		}
+
+		else if (ctx->revoke_statement() && ctx->revoke_statement()->ON() && ctx->revoke_statement()->permission_object()
```
Same as above

shalinilohia50 replied:
We are only supporting these privileges for this REVOKE on SCHEMA syntax. That's why we need this change in `exitSecurity_statement()`. Same as above.

KushaalShroff replied:
Same as above meant simplifying the condition :)

KushaalShroff commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -5764,12 +5815,129 @@ makeGrantdbStatement(TSqlParser::Security_statementContext *ctx)
+				}
+				if (prin->PUBLIC())
+				{
+					char *grantee_name = pstrdup("public");
```
We have used "public" a lot, we should rather have a macro

KushaalShroff commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -5764,12 +5815,129 @@ makeGrantdbStatement(TSqlParser::Security_statementContext *ctx)
+			if (ctx->revoke_statement()->permission_object()->full_object_name()->object_name)
+			{
+				schema_name = stripQuoteFromId(ctx->revoke_statement()->permission_object()->full_object_name()->object_name);
+				result->schema_name = pstrdup(downcase_truncate_identifier(schema_name.c_str(), schema_name.length(), true));
```
information schema rewrite not happening

KushaalShroff replied:
Where are we checking for restricted schemas/shared schemas?

shalinilohia50 replied:
Will do.

KushaalShroff commented on contrib/babelfishpg_tsql/src/tsqlIface.cpp:
```diff
@@ -5764,12 +5815,129 @@ makeGrantdbStatement(TSqlParser::Security_statementContext *ctx)
+				}
+			}
+			List *privilege_list = NIL;
+			for (auto perm: ctx->revoke_statement()->permissions()->permission())
+			{
+				auto single_perm = perm->single_permission();
+				if (single_perm->EXECUTE())
+					privilege_list = lappend(privilege_list, (void *)"execute");
+				if (single_perm->EXEC())
+					privilege_list = lappend(privilege_list, (void *)"execute");
+				if (single_perm->SELECT())
+					privilege_list = lappend(privilege_list, (void *)"select");
+				if (single_perm->INSERT())
+					privilege_list = lappend(privilege_list, (void *)"insert");
+				if (single_perm->UPDATE())
+					privilege_list = lappend(privilege_list, (void *)"update");
+				if (single_perm->DELETE())
+					privilege_list = lappend(privilege_list, (void *)"delete");
+				if (single_perm->REFERENCES())
+					privilege_list = lappend(privilege_list, (void *)"references");
```
As brought up multiple times in earlier reviews and earlier PRs, why cant we store the mask values directly? Why are we storing Char* and then later doing strcmp to fetch the mask values? :)

shalinilohia50 replied:
Why do we need to store mask values in `PLtsql_stmt_grantschema` instead of char*? 
Based on the privileges in the list, we generate the GRANT/ALTER PRIVILEGE queries. There will be multiple queries  for different privilege since the object type where it is applicable can be different.

If we do bitmask here and store the int value in the catalog, again we would need to get the list of privileges to generate GRANT schema subcommands in `gen_grantschema_subcmds`. 

IMO this seems better. Please suggest if there is a better way.

KushaalShroff replied:
You are already doing string comparison for gen_grantschema_subcmds. Storing a simple integer make MUCH more SENSE. Dont forget the benefit you get later during storing. :) 
You are then gonna compare atomically which is SUPER CHEAP in Time and space complexity, aint it? :)

shalinilohia50 replied:
Done.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql.h:
```diff
@@ -1067,6 +1068,20 @@ typedef struct PLtsql_stmt_fulltextindex
+	PLtsql_stmt_type cmd_type;
+	int			lineno;
+	bool		is_grant;
+	List		*privileges;		/* list of privileges */
```
Needs to be list of integers and not char*s

shalinilohia50 replied:
Why is list of char* a problem here? If we store list of integers, we would need to call `privilege_to_string` here to get the priv_name and generate subcommands.

KushaalShroff replied:
Check https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2215#discussion_r1455631001

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1978,3 +2011,154 @@ exec_alter_role_cmd(char *query_str, RoleSpec *role)
+}
+
+AclMode
+string_to_privilege(const char *privname)
```
We should store the masks upfront during parsing to avoid redundancy

shalinilohia50 replied:
Why is list of char* a problem here? If we store list of integers, we would need to call `privilege_to_string` here to get the priv_name and generate subcommands.

KushaalShroff replied:
CHECK https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2215#discussion_r1455631001

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1978,3 +2011,154 @@ exec_alter_role_cmd(char *query_str, RoleSpec *role)
+
+		/* do this step */
+		ProcessUtility(wrapper,
+					"(GRANT SCHEMA )",
```
Try using SPI

shalinilohia50 replied:
https://jira.rds.a2z.com/browse/BABEL-4698 Will be tracked here since it needs more understanding

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1978,3 +2011,154 @@ exec_alter_role_cmd(char *query_str, RoleSpec *role)
+ * Helper function to generate GRANT on SCHEMA subcommands.
+ */
+List
+*gen_grantschema_subcmds(const char *schema, const char *rolname, bool is_grant, bool with_grant_option, const char *privilege)
```
Explain why and how do we need to prevent SQL Injection?

Whatever done for "privilege" is it really a threat?

shalinilohia50 replied:
https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1848#discussion_r1358267697 Please check

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1978,3 +2011,154 @@ exec_alter_role_cmd(char *query_str, RoleSpec *role)
+					NULL);
+
+		/* make sure later steps can see the object created here */
+		CommandCounterIncrement();
```
Why?

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1033,16 +1033,49 @@ update_GrantStmt(Node *n, const char *object, const char *obj_schema, const char
+	ListCell *lc;
+
+	if (!IsA(stmt, AlterDefaultPrivilegesStmt))
+		ereport(ERROR, (errcode(ERRCODE_SYNTAX_ERROR), errmsg("query is not a AlterDefaultPrivilegesStmt")));
```
Why at this place?

shalinilohia50 replied:
Where else should we do it?  [References](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/0a3b2dca475894779afb8ebe302ba21bd411aae3/contrib/babelfishpg_tsql/src/pltsql_utils.c#L885-L957).

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql_utils.c:
```diff
@@ -1978,3 +2011,154 @@ exec_alter_role_cmd(char *query_str, RoleSpec *role)
+	else
+		update_AlterDefaultPrivilegesStmt(stmt, schema, rolname, privilege);
+
+	return stmt_list;
```
Query has not been freed (data)

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql.h:
```diff
@@ -2087,7 +2106,8 @@ extern void update_DropOwnedStmt(Node *n, List *role_list);
 extern void update_GrantRoleStmt(Node *n, List *privs, List *roles);
-extern void update_GrantStmt(Node *n, const char *object, const char *obj_schema, const char *grantee);
+extern void update_GrantStmt(Node *n, const char *object, const char *obj_schema, const char *grantee, const char *priv);
+extern void update_AlterDefaultPrivilegesStmt(Node *n, const char *object, const char *grantee, const char *priv);
```
Why is this function extern? I cant see its references in other files, are there?

shalinilohia50 replied:
Will make it static.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pltsql.h:
```diff
@@ -2049,6 +2066,8 @@ extern char *get_fulltext_index_name(Oid relid, const char *table_name);
 extern const char *gen_schema_name_for_fulltext_index(const char *schema_name);
 extern bool check_fulltext_exist(const char *schema_name, const char *table_name);
 extern bool is_unique_index(Oid relid, const char *index_name);
+extern List *gen_grantschema_subcmds(const char *schema, const char *rolname, bool is_grant, bool with_grant_option, const char *privilege);
```
Not being used in other files

shalinilohia50 replied:
Will make it static.

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+ * Removes all the rows corresponding to an OBJECT/SCHEMA from the catalog.
+ */
+void
+clean_up_object_from_bbf_schema(const char *schema_name,
```
Naming of function is confusing, be specific on which catalog are you cleaning

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+					F_NAMEEQ,
+					CStringGetDatum(schema_name));
+		scan = systable_beginscan(bbf_schema_rel,
+					get_bbf_schema_perms_idx_oid(),
```
Index is on around 5 columns, why are we trying to scan on index???

Lets discuss offline or update all scans accordingly

shalinilohia50 replied:
what do you suggest?

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+					F_NAMEEQ,
+					CStringGetDatum(object_name));
+		scan = systable_beginscan(bbf_schema_rel,
+					get_bbf_schema_perms_idx_oid(),
```
Index is on around 5 columns, why are we trying to scan on index???

shalinilohia50 replied:
What do you suggest?

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3606,6 +3792,24 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 }
 
+static void
+call_prev_ProcessUtility(PlannedStmt *pstmt,
```
Why do we need a new function for this? Your design to call call_prev_ProcessUtility before completing bbf seems odd, we are just doing this for updating catalogs post the grant.

We need to discuss this implementation offline, I am not keen in what has been implemented. Using SPI would have made more sense

shalinilohia50 replied:
https://jira.rds.a2z.com/browse/BABEL-4698 will be tracked here since it needs more understanding

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3290,7 +3299,17 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 
 					CommandCounterIncrement();
-
+					/* Grant ALL schema privileges to the user.*/
+					if (rolspec && strcmp(queryString, "(CREATE LOGICAL DATABASE )") != 0)
+					{
+						char *permissions[] = {"select", "insert", "update", "references", "delete", "execute"};
+						int i;
+						for (i = 0; i < 6; i++)
+						{
+							/* Execute the GRANT SCHEMA subcommands. */
+							exec_grantschema_subcmds(create_schema->schemaname, rolspec->rolename, true, false, permissions[i]);
+						}
```
During create schema is this expected?

shalinilohia50 replied:
Yes, this is expected for this type of CREATE schema syntax `create schema s1 authorization u1;`. We need to grant all the permissions on the schema to u1. We have tests added for this.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3580,16 +3600,182 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+										return;
+								}
+							}
+							break;
```
breaking from? Seems wrong

shalinilohia50 replied:
Removed it.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3580,16 +3600,182 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+								if (grant->is_grant)
+								{
+									call_prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params, queryEnv, dest, qc);
+									/* Don't add/update an entry, if the permission is granted on column list.*/
```
Is this supported?

shalinilohia50 replied:
We are not adding an entry in the catalog if permission is granted only on the column level. 
We have added few test cases with column level permission which works similar to SQL Server.

KushaalShroff replied:
Can you provide the document scoping column level permissions?

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3580,16 +3600,182 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+									 * 2. If permission on object exists, update the privilege in the catalog and revoke permission.
+									 */
+									if (entry_exists_in_bbf_schema(logical_schema, PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA, rol_spec->rolename))
+										has_schema_perms = true;
```
Is this really required?:)

Please refactor the call at the update to return, isnt it?

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -3580,16 +3600,182 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
+									return;
+							}
+						}
+						break;
```
Same here, why?

shalinilohia50 replied:
removed it.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3687,6 +3688,102 @@ get_insert_bulk_kilobytes_per_batch()
 	return insert_bulk_kilobytes_per_batch;
 }
 
+static int
```
Extern functions are static??

shalinilohia50 replied:
All these `exec_*` functions are implemented the same way if you see [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/0a3b2dca475894779afb8ebe302ba21bd411aae3/contrib/babelfishpg_tsql/src/pl_exec-2.c#L40-L48). I kept it consistent.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3687,6 +3688,102 @@ get_insert_bulk_kilobytes_per_batch()
+	char		*user = GetUserNameFromId(GetUserId(), false);
+	const char	*db_owner = get_owner_of_db(dbname);
+
+	login_is_db_owner = 0 == strncmp(login, db_owner, NAMEDATALEN);
```
NAMEDATALEN again, lets chat offline

shalinilohia50 replied:
Replaced strncmp with strcmp.

KushaalShroff commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3687,6 +3688,102 @@ get_insert_bulk_kilobytes_per_batch()
+
+	foreach(lc1, stmt->privileges)
+	{
+		char	*priv_name = (char *) lfirst(lc1);
```
Update this after changing list to integer

shalinilohia50 replied:
Why is list of char* a problem here? If we store list of integers, we would need to call `privilege_to_string` here to get the priv_name and generate subcommands.

KushaalShroff replied:
CHeck: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2215#discussion_r1455631001

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+ * Checks if a particular catalog entry exists in BABELFISH_SCHEMA_PERMISSIONS.
+ */
+bool
+entry_exists_in_bbf_schema(const char *schema_name,
```
Ambiguous and confusing Naming of function

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+	table_close(bbf_schema_rel, RowExclusiveLock);
+
+	/* Advance cmd counter to make the insert visible */
+	CommandCounterIncrement();
```
Over-use of commandCounterIncrement, why would we require the tuple to be visible just after? Wont it be just a single grant statement in that query? Also in code flow we will CommandCounterIncrement more than once, can you validate?

shalinilohia50 replied:
In case of grant statement with multiple grantees, this will be executed that number of times. 
`grant select, execute on schema::s1 to u1, u2` -> There will be multiple grant statements for different privileges as well, since select and execute applies to different categories of object and we form grant statements based on that.

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+				CStringGetDatum(grantee));
+
+	scan = table_beginscan_catalog(bbf_schema_rel, 5, scanKey);
+	tuple_bbf_schema = heap_getnext(scan, ForwardScanDirection);
```
Raised in previous PRs: Why are we not consistent with sys calls and heap calls

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+		return;
+
+	/* Get existing privilege of an object. */
+	old_permission = get_privilege_of_object(schema_name, object_name, grantee, object_type);
```
redundancy again: looking up same table twice. Why cant we do it during a single lookup?

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+		return false;
+
+	/* Immediately return false, if SCHEMA name is NULL. */
+	if (schema_name == NULL)
```
We must throw errors/return false for system/shared schemas

shalinilohia50 replied:
Will do.

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+ * Checks if a particular schema has any SCHEMA level permission granted to any user.
+ */
+bool
+permission_on_schema_exists(const char *schema_name,
```
Do we really require a new function for that?

Why cant we unionise the logic into entry_exists_in_bbf_schema conditionally?

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+ * Add an entry to BABELFISH_SCHEMA_PERMISSIONS table, if it doesn't exist already.
+ * If exists, updates the PERMISSION column in the table.
+ */
+void
+add_or_update_object_in_bbf_schema(const char *schema_name,
+				const char *object_name,
+				int new_permission,
+				const char *grantee,
+				const char *object_type,
+				bool is_grant)
+{
+	if (!entry_exists_in_bbf_schema(schema_name, object_name, grantee))
+		add_entry_to_bbf_schema_perms(schema_name, object_name, new_permission, grantee, object_type);
+	else
+		update_privileges_of_object(schema_name, object_name, new_permission, grantee, object_type, is_grant);
+}
```
Can be a macro

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+			initStringInfo(&query);
+			/* Check if the permission to be REVOKED on SCHEMA exists on the OBJECT. */
+
+			if (permission == (current_permission & permission))
```
Previous PR's this comment not addressed: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2044#discussion_r1401052015

KushaalShroff commented:
Linking another closed PR reviewed in past: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/2044

lohia-shalini committed:
replace NAME datatype with nvarchar

lohia-shalini committed:
Emmpty commit

lohia-shalini committed:
Remove permission_on_schema_exists()

lohia-shalini committed:
Simplify code

lohia-shalini committed:
Return immediately for shared schema

lohia-shalini committed:
stored integer instead of list of char* for permissions

lohia-shalini committed:
Get rid of clear_rewritten_query_fragment

shalinilohia50 committed:
Merge branch 'BABEL_3_X_DEV' into grant-on-schema

Deepesh125 commented on test/JDBC/expected/GRANT_SCHEMA-vu-prepare.out:
```diff
@@ -0,0 +1,74 @@
+create proc grant_schema_s1.grant_schema_p2 as select 2;
+go
+
+CREATE FUNCTION grant_schema_s1.grant_schema_f1() RETURNS INT AS BEGIN RETURN (SELECT COUNT(*) FROM sys.objects) END
```
count(*) is probably not right choice for testing as it may vary

shalinilohia50 replied:
Updated this test.

Deepesh125 commented on test/JDBC/expected/GRANT_SCHEMA-vu-verify.out:
```diff
@@ -0,0 +1,291 @@
+-- tsql user=grant_schema_l1 password=12345678
```
can you add negative test cases when permission is outside of select or execute

shalinilohia50 replied:
Done.

Deepesh125 commented on test/JDBC/jdbc_schedule:
```diff
@@ -39,6 +39,9 @@ ignore#!#sys-types-before-dep-vu-cleanup
 ignore#!#sys-table_types-before-dep-vu-prepare
 ignore#!#sys-table_types-before-dep-vu-verify
 ignore#!#sys-table_types-before-dep-vu-cleanup
+ignore#!#GRANT_SCHEMA-vu-prepare
+ignore#!#GRANT_SCHEMA-vu-verify
+ignore#!#GRANT_SCHEMA-vu-cleanup
```
why do we need to ignore in latest release?

shalinilohia50 replied:
For the objects where the permission is granted on the older version, and REVOKE on schema performed on the newer version then permission will be revoked from those older objects. These tests are specifically for Upgrade scenario and it will fail in the current release.

Deepesh125 commented on test/JDBC/expected/GRANT_SCHEMA.out:
```diff
@@ -0,0 +1,845 @@
+go
+
+-- Mixed case
+grant Execute on Babel_4344_F1 to babel_4344_u1;
+go
+grant execute on babel_4344_s1.babel_4344_f1 to babel_4344_u1;
+go
```
user should also be in mixed cases. Where/how are we verifying if this works as expected?

Deepesh125 replied:
schema name should also be in mixed cases

Deepesh125 replied:
grant select on schema in one case and then use different case during revoke or different case during select. Same should be done at object level

shalinilohia50 replied:
All these tests are added. Where the grant on schema/object is in one case and revoke/select/execute is in a different case.

Deepesh125 reviewed (commented):
1. I still cant find much coverage in terms of mixed cases, quoted identifiers and languages --- ideally we need to have mix cases during grant / revoke and verification of all these cases. 
2. Another thing I am wondering is about concurrency, How updates being done in one session would be visible in concurrent session. How can we test this?
3. Another use cases missing is, repeated permission grant on same objects and checking the catalog consistency.
4. ALTER object use cases are totally missing

lohia-shalini committed:
Add mixed case tests

shalinilohia50 commented:
> 1. I still cant find much coverage in terms of mixed cases, quoted identifiers and languages --- ideally we need to have mix cases during grant / revoke and verification of all these cases.
> 2. Another thing I am wondering is about concurrency, How updates being done in one session would be visible in concurrent session. How can we test this?
> 3. Another use cases missing is, repeated permission grant on same objects and checking the catalog consistency.
> 4. ALTER object use cases are totally missing

1. I have tests for mixed cases. Will add more for quoted identifiers and languages.
2. I don't think we can do that in JDBC framework, though I have tested it locally and validated.
3. Added this test case.
4. Can you give some examples? Do you mean sp_rename? If yes, it will be handled in a separate PR on top of this change.

lohia-shalini committed:
Add different language tests

lohia-shalini committed:
add tests where objects are in quotes

shalinilohia50 commented:
> > 1. I still cant find much coverage in terms of mixed cases, quoted identifiers and languages --- ideally we need to have mix cases during grant / revoke and verification of all these cases.
> > 2. Another thing I am wondering is about concurrency, How updates being done in one session would be visible in concurrent session. How can we test this?
> > 3. Another use cases missing is, repeated permission grant on same objects and checking the catalog consistency.
> > 4. ALTER object use cases are totally missing
> 
> 1. I have tests for mixed cases. Will add more for quoted identifiers and languages.
> 2. I don't think we can do that in JDBC framework, though I have tested it locally and validated.
> 3. Added this test case.
> 4. Can you give some examples? Do you mean sp_rename? If yes, it will be handled in a separate PR on top of this change.

Added all these tests. Please check.

Deepesh125 commented:
As discussed offline, we would like to have a tests cases with quoted identifiers at various levels like schame name, object name and role name. And mix it with case-insensitivity kind of testing. Please make sure that we use quoted identifier which can be used within regular identifiers, for example, white space. 

And lets have one or two test cases for multibyte chars (such as japanese or chinese). Already added tests such as arabic looks good but it is single byte. 

And make sure that we add few test cases where identifiers length gets breached (>64bytes?)

Example that is in my mind --
```
1> create schema "select 1" 
2> go
1> create table "select 1"."select 2" (a int)
2> go
1> grant select on schema::"selEct 1" to Dipesh
2> go
1> -- verify dipesh is able to access schema 1 
2> revoke select on schema::"SELECT 1" FROM "dipesh"
3> go
```

lohia-shalini committed:
Add japanese character, tests with quoted identifiers

lohia-shalini committed:
Add test cases from sql injection perspective

lohia-shalini committed:
handle special roles like db_owner, sys, information_schema

Deepesh125 commented on contrib/babelfishpg_tsql/src/pl_exec-2.c:
```diff
@@ -3729,19 +3729,32 @@ exec_stmt_grantschema(PLtsql_execstate *estate, PLtsql_stmt_grantschema *stmt)
+		if (!is_public && !OidIsValid(role_oid))
+		{
+			/* sys or information_schema roles should throw an error. */
+			if ((strcmp(grantee_name, "sys") == 0) || (strcmp(grantee_name, "information_schema") == 0))
```
information_schema_tsql also?

shalinilohia50 replied:
why?
`information_schema_tsql` is not a predefined SQL/database roles.

shalinilohia50 replied:
that's for schema, grantee name is not rewritten to `information_schema_tsql`. Refer the catalog `sys.database_principals`
